inbm/dispatcher-agent/tests/ai/snapshot_genai_unit_test.py

```diff
@@ -0,0 +1,1099 @@
+from dispatcher.sota.rebooter import Rebooter  # Changed to absolute import
+from unittest.mock import Mock, patch
+from dispatcher.sota.sota_error import SotaError  # Import SotaError
+from unittest.mock import patch, MagicMock
+from dispatcher.sota.snapshot import mender_commit_command
+from dispatcher.sota.snapshot import Snapshot, Rebooter  # Import Rebooter
+from dispatcher.sota.snapshot import Snapshot, Rebooter
+from dispatcher.sota.snapshot import Snapshot, Trtl  # Import Trtl here
+from dispatcher.sota.snapshot import Snapshot
+from dispatcher.sota.snapshot import WindowsSnapshot
+from inbm_lib.trtl import Trtl
+from inbm_lib.trtl import Trtl  # Add this line to import Trtl
+from dispatcher.sota.snapshot import YoctoSnapshot
+from dispatcher.dispatcher_exception import DispatcherException
+from dispatcher.sota.sota_error import SotaError  # Added this line
+from dispatcher.sota.snapshot import YoctoSnapshot, SotaError  # Import SotaError
+
+#DO NOT DELETE THIS LINE - TestDebianBasedSnapshotCommit
+'''
+ADD HUMAN FEEDBACK BELOW:
+
+'''
+class TestDebianBasedSnapshotCommit:
+    @pytest.mark.parametrize('snap_num, delete_snapshot_return, expected_output', [
+        # Test when snap_num is None
+        (None, (1, 'snap_num is None'), 1),
+        # Test when snap_num is 0
+        (0, (0, 'snap_num is 0 (dummy snapshot); no need to delete'), 0),
+        # Test when snap_num is valid and delete_snapshot is successful
+        ('123', (0, None), 0),
+        # Test when snap_num is valid but delete_snapshot fails
+        ('123', (1, 'Error message'), 1)
+    ])
+    def test_commit(self, snap_num, delete_snapshot_return, expected_output):
+        # Mock the Trtl instance
+        trtl = MagicMock()
+        trtl.delete_snapshot.return_value = delete_snapshot_return
+
+        # Mock the DispatcherBroker instance
+        dispatcher_broker = MagicMock()
+
+        # Create a DebianBasedSnapshot instance with the mocked Trtl and DispatcherBroker
+        snapshot = DebianBasedSnapshot(trtl, 'sota_cmd', dispatcher_broker, snap_num, False, False)
+
+        # Call the commit method and check the output
+        output = snapshot.commit()
+        assert output == expected_output
+
+        # Check that the delete_snapshot method was called with the correct arguments
+        if snap_num is not None and snap_num != 0:
+            trtl.delete_snapshot.assert_called_once_with(snap_num)
+
+        # Check that the telemetry method was called with the correct arguments
+        if delete_snapshot_return[0] == 0:
+            dispatcher_broker.telemetry.assert_called_once_with("Snapshot cleanup succeeded")
+        else:
+            dispatcher_broker.telemetry.assert_called_once_with(f"SOTA snapshot delete failed: {delete_snapshot_return[1]}")
+
+#DO NOT DELETE THIS LINE - TestDebianBasedSnapshotRecover
+'''
+ADD HUMAN FEEDBACK BELOW:
+
+'''
+class TestDebianBasedSnapshotRecover:
```

Comment:
To go back to Pull Request Summary table, Click on the [link](https://github.com/intel/intel-inb-manageability/pull/489#)

<h3>Class: DebianBasedSnapshot |Function: recover</h3>

[Source Code Link](https://github.com/intel/intel-inb-manageability/blob/NEX_MAXIMUS_UT/test-specific-unit-test/inbm/dispatcher-agent/dispatcher/sota/snapshot.py?plain=1#L218)

[Unit Test Case Code Link](https://github.com/intel/intel-inb-manageability/blob/NEX_MAXIMUS_UT/test-specific-unit-test/inbm/dispatcher-agent/tests/ai/snapshot_genai_unit_test.py?plain=1#L72)

<h4>Test Case Scenarios:</h4>
The unit test `test_recover` in the `TestDebianBasedSnapshotRecover` class generates six test cases based on the parameters provided to the `@pytest.mark.parametrize` decorator. Here are the test cases along with their explanations:

1. Test Case: `test_recover["0", True, 5, 1]`
   - Explanation: This test case checks the `recover` method when `snap_num` is "0", `reboot_device` is True, `time_to_wait_before_reboot` is 5 seconds, and the expected number of calls to the `reboot` method is 1. Since `snap_num` is "0", the `_rollback_and_delete_snap` method should not be called, and the system should reboot once.

2. Test Case: `test_recover["0", False, 5, 0]`
   - Explanation: This test case checks the `recover` method when `snap_num` is "0", `reboot_device` is False, `time_to_wait_before_reboot` is 5 seconds, and the expected number of calls to the `reboot` method is 0. Since `snap_num` is "0" and `reboot_device` is False, neither the `_rollback_and_delete_snap` method nor the `reboot` method should be called.

3. Test Case: `test_recover["123", True, 5, 1]`
   - Explanation: This test case checks the `recover` method when `snap_num` is "123", `reboot_device` is True, `time_to_wait_before_reboot` is 5 seconds, and the expected number of calls to the `reboot` method is 1. Since `snap_num` is not "0", the `_rollback_and_delete_snap` method should be called once, and the system should reboot once.

4. Test Case: `test_recover["123", False, 5, 1]`
   - Explanation: This test case checks the `recover` method when `snap_num` is "123", `reboot_device` is False, `time_to_wait_before_reboot` is 5 seconds, and the expected number of calls to the `reboot` method is 1. Even though `reboot_device` is False, since `snap_num` is not "0", the `_rollback_and_delete_snap` method should be called once, and the system should reboot once.

5. Test Case: `test_recover[None, True, 5, 1]`
   - Explanation: This test case checks the `recover` method when `snap_num` is None, `reboot_device` is True, `time_to_wait_before_reboot` is 5 seconds, and the expected number of calls to the `reboot` method is 1. Since `snap_num` is None, the `_rollback_and_delete_snap` method should not be called, and the system should reboot once.

6. Test Case: `test_recover[None, False, 5, 0]`
   - Explanation: This test case checks the `recover` method when `snap_num` is None, `reboot_device` is False, `time_to_wait_before_reboot` is 5 seconds, and the expected number of calls to the `reboot` method is 0. Since `snap_num` is None and `reboot_device` is False, neither the `_rollback_and_delete_snap` method nor the `reboot` method should be called.

<h4>Short summary about the testcases:</h4>
No failed cases

<h4>Test Execution Results:</h4>

| Total Cases    | 6  |
| -------------- | -- |
| Passed Cases   | 6  |
| Failed Cases   | 0  |



To go back to Pull Request Summary table, Click on the [link](https://github.com/intel/intel-inb-manageability/pull/489#)